### PR TITLE
[CALCITE-2596] AssertionError: Not-null boxed primitive should come back as primitive when translating correlated variables (Stamatis Zampetakis)

### PR DESCRIPTION
**Pull request by zabetak**
Apply nullAs policy when translating expressions with correlated variables.

### CODE REVIEW

**Comment by zabetak** [comment]
This PR will create an easy conflict with #762.

**Comment by vlsi** [comment]
I wonder if the root cause is that `org.apache.calcite.adapter.enumerable.JavaRowFormat#field` ignores `Type fieldType` parameter

**Comment by zabetak** [comment]
I am not sure if it comes from the fieldType parameter. However, I think that applying the NullAs policy to the expression is important otherwise we are going to end up with compile errors in the generated code.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 1]
```diff
@@ -667,7 +675,10 @@ private Expression translate0(RexNode expr, RexImpTable.NullAs nullAs,
       }
       InputGetter getter =
           correlates.apply(((RexCorrelVariable) target).getName());
-      return getter.field(list, fieldAccess.getField().getIndex(), storageType);
+      Expression y = getter.field(list, fieldIndex, storageType);
+      Expression input = list.append("corInp" + fieldIndex + "_", y); // safe to share
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: This is not required.

Does it really add any value?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 2]
```diff
@@ -569,6 +569,18 @@ Expression translateCast(
     return scaleIntervalToNumber(sourceType, targetType, convert);
   }
 
+  private Expression handleNullUnboxingIfNecessary(
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: Technically speaking this is not just `unboxing`.

`org.apache.calcite.adapter.enumerable.RexImpTable.NullAs#IS_NOT_NULL` means the expression `x` should be translated to `x is not null` kind of format.

In other words, `translate(x, NullAs.IS_NOT_NULL)` is equivalent to `translate(isNotNull(x), NOT_POSSIBLE)`, so `handleNull` sometimes adds `==null` or `!=null` kind of expressions.

Would you please rename accordingly?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 3]
```diff
@@ -653,8 +660,9 @@ private Expression translate0(RexNode expr, RexImpTable.NullAs nullAs,
     case CORREL_VARIABLE:
       throw new RuntimeException("Cannot translate " + expr + ". Correlated"
           + " variables should always be referenced by field access");
-    case FIELD_ACCESS:
+    case FIELD_ACCESS: {
       RexFieldAccess fieldAccess = (RexFieldAccess) expr;
+      int fieldIndex = fieldAccess.getField().getIndex();
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: This is not required provided you don't add `"corInp" + fieldIndex` below

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 2 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: I totally agree. The way I am reading the method is handleNull, unboxing if necessary but if that creates confusion I will try to think of a better name.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 3 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: The way the code is on the master branch it is needed even if "corInp" + fieldIndex is removed.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 2 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: I was reading it as "handle NullUnboxing if necessary"

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 3 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: That's fine then

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 1 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: I have it like this for two reasons:
- consistency with the existing code 639-641 (case INPUT_REF);
- readability of generated code.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java (hunk 2 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: What about the following name "handleNullAndUnboxIfNecessary"?

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: I guess that would work.
`handleNull` would work as well.

Comment (zabetak) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java]: Done

**Comment by zabetak** [comment]
The conflicts are solved and I resolved the comments from @vlsi. 

Do you have more comments? Should I rebase now?

**Pull request closed without merging**